### PR TITLE
eni: refactor(metrics) IPAM related metrics

### LINKED ISSUE #20384: ipam metrics:  number of interfaces with addresses available is 0 in no ip require and no ip release when opeartor restart
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

1.   If all node no ip require and no ip release in ipam eni mode, the metrics `cilium_operator_ipam_available` is zero.
2. The relate code is https://github.com/cilium/cilium/blob/e872b2c34aed1e99b9672d291ed9ecc7d5b41af2/pkg/ipam/node.go#L564-L608
3. If `stats.NeededIPs == 0`, it is not run the `n.ops.PrepareIPAllocation`, so no set the `n.stats.RemainingInterfaces = a.allocation.AvailableInterfaces` . 
4. And in `resyncNode`  the `stats.remainingInterfaces` is 0
https://github.com/cilium/cilium/blob/e872b2c34aed1e99b9672d291ed9ecc7d5b41af2/pkg/ipam/node_manager.go#L396-L414

### Cilium Version

[v1.12.0-rc3](https://github.com/cilium/cilium/releases/tag/v1.12.0-rc3)

### Kernel Version

no

### Kubernetes Version

no

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #20704: eni: need redesign eni metric, many wrong and ambiguities metrics
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Many wrong metrics that has no label value or metrics name is ambiguities in eni metrics.
such as:
1.  Failed to release ip that call `IncAllocationAttempt` in `handleIPRelease`. `IncAllocationAttempt` is mean "Number of interfaces allocated", it not represent an error in releasing IP. https://github.com/cilium/cilium/blob/78d075d1bedf7f6e3890dd2a9fab20383201f3c4/pkg/ipam/node.go#L776-L806
2. It to call `IncAllocationAttempt` to add `cilium_operator_ipam_interface_creation_ops` value  when success call `AllocateIPs` in `handleIPAllocation`. it is difficult to distinguish whether the network card is created successfully or the IP is allocated successfully. https://github.com/cilium/cilium/blob/78d075d1bedf7f6e3890dd2a9fab20383201f3c4/pkg/ipam/node.go#L821-L836
3. https://github.com/cilium/cilium/issues/20384
4. https://github.com/cilium/cilium/issues/20431

### Cilium Version

master https://github.com/cilium/cilium/tree/78d075d1bedf7f6e3890dd2a9fab20383201f3c4

### Kernel Version

none

### Kubernetes Version

none

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by wu0407**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

1. fix `cilium_operator_ipam_available` is zero when all node are no need ip.
2. refactor ipam allocation and release metrics.
  2.1 add `cilium_operator_allocation_duration_seconds` and `cilium_operator_release_duration_seconds`
3. rename metrics name
    3.1 `cilium_operator_ipam_available` to `cilium_operator_ipam_available_interfaces`
    3.2 `cilium_operator_ipam_allocation_ops` to `cilium_operator_ipam_ip_allocation_ops`
    3.3 `cilium_operator_ipam_release_ops` to `cilium_operator_ipam_ip_release_ops`

Fixes: #20704
Fixes: #20384

```release-note
Add new ENI IPAM metrics for allocation, release
```

### CODE REVIEW

File: pkg/aws/eni/node.go
Comment (gandro) [inline pkg/aws/eni/node.go]: This block is copied over directly from `PrepareIPAllocation`, correct? Can we move that logic into a separate function somehow? Otherwise we run into the risk that if one place is changed, that the other one will not, and thus lead to inconsistencies.

File: pkg/azure/ipam/node.go
Comment (gandro) [inline pkg/azure/ipam/node.go:180]: Similarly to the comment above, let's move this into a separate function. Otherwise the logic will diverge over time.

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: Could you add a comment to the source as to why this has been moved down? It's not clear to me from context.

Also, I'm nervous about the fact that we now write `n.resource = resource` without creating a copy first. That could introduce data races. Can you keep the deep copy at the top of the function, and only move the `UpdatedNode` call?

**Review by gandro — CHANGES_REQUESTED** [review]
Great improvements, thank you! A few minor things which I believe should be addressed before we can merge this.

File: pkg/ipam/node.go
Comment (wu0407) [reply pkg/ipam/node.go]: this my mistake,  I add comment for that, update n.resource before n.ops.UpdatedNode.

File: pkg/azure/ipam/node.go
Comment (wu0407) [reply pkg/azure/ipam/node.go:180]: Is mean error  handle need into a separate function? I move available logic in separate function.
``` go
availableOnInterface, available := isAvailableInterface(requiredIfaceName, iface, scopedLog)
		if !available {
			return nil
		}
```

File: pkg/aws/eni/node.go
Comment (wu0407) [reply pkg/aws/eni/node.go]: resloved in https://github.com/cilium/cilium/compare/ffe5a341ef6bd7ae923aaec4bcd6bac2cc24bfa1..2ef0193d8fb3a85b517ad6d4704022f94016fed6

**Comment by tommyp1ckles** [comment]
Thanks for the PR!

Could you split this PR into two commits such that both issues being fixed are separate commits. I'm finding it hard to follow what changes belong to what fix.

**Review by christarazi — COMMENTED** [review]
Thanks the PR, +1 to Tom's feedback about splitting commits.

**Comment by wu0407** [comment]
> Thanks the PR, +1 to Tom's feedback about splitting commits.

it is done @christarazi @tommyp1ckles

File: pkg/azure/ipam/node.go
Comment (tommyp1ckles) [inline pkg/azure/ipam/node.go]:
```suggestion
			}).Debug("Not considering interface as available since it does not match the required name")
```

File: pkg/azure/ipam/node.go
Comment (tommyp1ckles) [inline pkg/azure/ipam/node.go]:
```suggestion
	}).Debug("Considering interface as available")
```

**Review by tommyp1ckles — APPROVED** [review]
Azure changes look fine, thanks @wu0407

**Comment by wu0407** [comment]
Is need fix this issue https://github.com/cilium/cilium/issues/20815 in this pr ?

File: pkg/azure/ipam/node.go
Comment (gandro) [reply pkg/azure/ipam/node.go:180]: Apologies, I missed that. This is good as is. Minor nit: `remainAvailableENIsCount` is not a good variable name here, since Azure network interfaces are not ENIs. Maybe use  `remainingAvailableInterfaceCount` instead?

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: Similarly as above, let's call this ` remainingAvailableInterfaceCount` to keep it generic

File: pkg/ipam/node_manager.go
Comment (gandro) [inline pkg/ipam/node_manager.go]: Similarly as above, let's call this remainingAvailableInterfaceCount to keep it generic

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: nit, better grammar:

```suggestion
	// of performing a complete n.recalculate() execution when the nodeManager performs a resync
	// and the operator is starting up.
```

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]:
```suggestion
	// This best effort to solve the issue where the metrics `cilium_operator_ipam_available_interfaces`
	// only contains a part of the nodes at operator startup. It will be set to the correct
	// value after the next period where nodeManager performs a resync.
```

**Review by gandro — COMMENTED** [review]
Thanks for addressing the feedback, some more nits

**Comment by gandro** [comment]
> Is need fix this issue #20815 in this pr ? @gandro @christarazi

Ideally, let's keep it in a separate PR

**Comment by wu0407** [comment]
> Thanks for addressing the feedback, some more nits

it is done @gandro

File: pkg/aws/eni/node.go
Comment (gandro) [inline pkg/aws/eni/node.go]: nit:

```suggestion
	return effectiveLimits
```

File: pkg/azure/ipam/node.go
Comment (gandro) [inline pkg/azure/ipam/node.go]: nit:

```suggestion
// isAvailableInterface returns whether interface is available and the number of available IPs to allocate in interface
```

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: nit:

```suggestion
	// This is best effort to solve the issue where the metrics `cilium_operator_ipam_available_interfaces`
```

**Review by gandro — APPROVED** [review]
Thanks, this looks good to me. Three more minor non-blocking nits.

I will be out of office for the next two weeks, so I'll rely on the other reviewers on this PR to drive this to completion.

**Review by qmonnet — APPROVED** [review]
Looks good, thanks!

File: pkg/alibabacloud/eni/node.go
Comment (christarazi) [inline pkg/alibabacloud/eni/node.go]: Error strings should be lower-cased.

File: pkg/ipam/node_manager.go
Comment (christarazi) [inline pkg/ipam/node_manager.go]: I think instead of having named returns here, could you document what the return values are meant to be in the godoc.

Comment (wu0407) [reply pkg/ipam/node_manager.go]: it done

File: pkg/aws/eni/node.go
Comment (christarazi) [inline pkg/aws/eni/node.go]: Still seeing error strings as upper-cased. Should be lower-cased

File: pkg/ipam/node_manager.go
Comment (christarazi) [reply pkg/ipam/node_manager.go]: Could you remove the named returns in the interface declaration? There's no need for them there, just the godoc changes are fine.

Comment (wu0407) [reply pkg/ipam/node_manager.go]: Ok, i remove the named returns

File: pkg/aws/eni/node.go
Comment (wu0407) [reply pkg/aws/eni/node.go]: replace all the "Unable to determine limits" to lower

File: Documentation/operations/upgrade.rst
Comment (joestringer) [inline Documentation/operations/upgrade.rst]: nit: Can you please move this up into a new section for v1.13 upgrade notes? It doesn't make sense to state this in the v1.11 section. If there is not the existing `ADded metrics` titles etc. in the v1.13 section above, please create it.

**Review by joestringer — CHANGES_REQUESTED** [review]

**Comment by joestringer** [comment]
The Travis / go lint failures look related to a breakage on master. We've fixed that now, so if you rebase then these tests should pass. Sorry about the noise.

**Review by christarazi — APPROVED** [review]
LGTM on the ENI side, pending Joe's comments.

File: Documentation/operations/upgrade.rst
Comment (wu0407) [reply Documentation/operations/upgrade.rst]: current no `1.13 Upgrade Notes` in [Documentation/operations/upgrade.rst](https://github.com/cilium/cilium/pull/20755/files/73a4efa0a2fb4eb2d60aba8e2d533955e38423f7#diff-76c7d3d61bd3c6afeb8e936896c7a01046d530840abf499fd118e88e75d1a504), i will add it.

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
Just a heads up there will be a minor conflict vs. #20255. Please update the PR to specify which of these ipam metrics are enabled by default, to fit with the new table format in that conflicting docs page.

**Comment by wu0407** [comment]
> 
How to resolve the conficts?
need merge master or only change conflicting docs page

**Comment by wu0407** [comment]
> > 
> 
> How to resolve the conficts? need merge master or only change conflicting docs page @joestringer

rebase done

**Comment by tklauser** [comment]
test-1.16-4.9 failed on known flake #20288, all other tests are green and approvals are in. Merging.

**Pull request merged**
